OpenRA.Game/Graphics/WorldRenderer.cs

```diff
@@ -126,14 +126,16 @@ List<IFinalizedRenderable> GenerateRenderables()
 			return renderables;
 		}
 
-		List<IFinalizedRenderable> GenerateOverlayRenderables()
+		List<IFinalizedRenderable> GenerateOverlayRenderables(IEnumerable<Actor> actorsInBox)
 		{
-			var aboveShroud = World.ActorsWithTrait<IRenderAboveShroud>().Where(a => a.Actor.IsInWorld && !a.Actor.Disposed)
-				.SelectMany(a => a.Trait.RenderAboveShroud(a.Actor, this));
+			var aboveShroud = World.ActorsWithTrait<IRenderAboveShroud>()
+				.Where(a => a.Actor.IsInWorld && !a.Actor.Disposed && (!a.Trait.SpatiallyPartitionable || actorsInBox.Contains(a.Actor)))
+					.SelectMany(a => a.Trait.RenderAboveShroud(a.Actor, this));
 
-			var aboveShroudSelected = World.Selection.Actors.Where(a => !a.Disposed)
-				.SelectMany(a => a.TraitsImplementing<IRenderAboveShroudWhenSelected>()
-					.SelectMany(t => t.RenderAboveShroud(a, this)));
+			var aboveShroudSelected = World.ActorsWithTrait<IRenderAboveShroudWhenSelected>()
```

Comment:
Most actors have at least one `IRenderAboveShroudWhenSelected` trait, so this change introduces a lot more work.  Can we please keep this enumerating over the much smaller `World.Selection.Actors.SelectMany()` collection?